### PR TITLE
Implement command to migrate u2f registrations to webauthn

### LINKED ISSUE #6: twofactor_u2f compatibility
Add compatibility to twofactor_u2f keys

### PR DESCRIPTION
**Pull request by ChristophWurst**
- [x] Upstream patch (https://github.com/web-auth/webauthn-framework/pull/208)
- [x] Wait for next release of web-auth/webauthn-lib (probably v3.3.11)
- [x] Copy mapper and entity from u2f app
- [x] Delete old registrations?
- [x] Catch exceptions during migration (see my TODO in the command class)?

Fixes #6

### CODE REVIEW

**Comment by st3iny** [comment]
Upstream fixes to make login work at https://github.com/web-auth/webauthn-framework/pull/207.

**Comment by levihb** [comment]
If all of our users have twofactor_u2f, can this migrate them all to this?

**Comment by ChristophWurst** [comment]
yes, that is the goal

**Comment by st3iny** [comment]
The other PR (I linked) was merged and fixes our issues too. The migration is working fine when tested with the the latest git version of web-auth/webauthn-framework.

@ChristophWurst Additionally, it will now be possible to distinguish old U2F from regular Webauthn credentials in the future because we don't have to convert public keys anymore. The other PR was a bit more elegant than my proposal. Just in case something goes wrong.

File: lib/Service/WebauthnManager.php [hunk 1]
```diff
@@ -89,7 +89,7 @@ public function __construct(
 		ISession $session,
 		WebauthnPublicKeyCredentialSourceRepository $repository,
 		PublicKeyCredentialEntityMapper $mapper,
-		IEventDispatcher $eventDispatcher,
+		IEventDispatcher $eventDispatcher
```
Comment (st3iny) [inline lib/Service/WebauthnManager.php:92]: Is not strictly related but fixes compatibility with PHP <= 7.4.

**Comment by ChristophWurst** [comment]
Bildschirmfoto von 2022-02-23 11-49-10

:smiling_face_with_tear:

**Comment by ChristophWurst** [comment]
@nickvergessen I'll try to debug this a bit. If there isn't anything obvious to fix I'd like you to also give this a go with your authentication device. I'll let you know.

File: lib/Command/MigrateU2F.php [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+		return 0;
+	}
+
+	private function migrateUser(IUser $user, OutputInterface $output): void {
+		$output->writeln('Migrating devices of user ' . $user->getUID());
+		$registrations = $this->u2fMapper->findRegistrations($user);
+		foreach ($registrations as $registration) {
+			$name = $registration->getName() . ' (U2F)';
+			if (strlen($name) > 64) {
+				$dots = '...';
```
Comment (nickvergessen) [inline lib/Command/MigrateU2F.php:116]:
```suggestion
				$dots = '…';
```

File: lib/Service/U2FMigrator.php [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+		// required. It is only required in super high security contexts
+		// as it complicates the whole second factor UX.
+		// Ref https://developers.yubico.com/WebAuthn/WebAuthn_Developer_Guide/Attestation.html
+		// TL;DR: Attestation ensures that only specific/approved devices with
+		// signed certificates can be used. Like TLS but for Webauthn devices.
+		$attestationType = AttestationStatement::TYPE_NONE;
+		$trustPath = new EmptyTrustPath();
+
+		// Credential ID is the same as the legacy key handle
+		$credentialId = $this->base64_urlsafe_decode($registration->getKeyHandle());
```
Comment (nickvergessen) [inline lib/Service/U2FMigrator.php:61]: This somehow does strange things?
`oc_twofactor_u2f_registrations.key_handle` start with `g0vrE-` for me and
`oc_twofactor_webauthn_registrations.public_key_credential_id` starts with `g0vrE+` 

And webauthn is not working afterwards.
is this expected?

**Comment by nickvergessen** [comment]
I can not login after migrate-u2f

File: lib/Service/U2FMigrator.php (hunk 3 above)
Comment (ChristophWurst) [reply lib/Service/U2FMigrator.php:61]: I think I'm running into the same. Will debug.

Comment (nickvergessen) [reply lib/Service/U2FMigrator.php:61]: Also the value ends one time on `OQ` and one time on `OQ==`

**Comment by ChristophWurst** [comment]
I realized that https://github.com/web-auth/webauthn-framework/pull/208 is still not released and we are using a broken webauthn lib.

**Comment by ChristophWurst** [comment]
@nickvergessen now it works for me.

**Comment by ChristophWurst** [comment]
> Delete old registrations?

I tend to say we delete the u2f registration by default but maybe offer a flag to keep them. That would mean the U2F app is not active anymore for any user after the migration and can be safely disabled. @nickvergessen what do you think?

**Comment by nickvergessen** [comment]
> now it works for me.

Yes, same here

> I tend to say we delete the u2f registration by default but maybe offer a flag to keep them. That would mean the U2F app is not active anymore for any user after the migration and can be safely disabled.

I would accept both, either optionally deleting or optionally keeping

**Review by nickvergessen — APPROVED** [review]

**Pull request merged**